File: 2024/java/senior/Senior4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Senior4 {
    static List<List<int[]>> roads = new ArrayList<>();
    static String[] colors;
    static Set<Integer> visited = new HashSet<>();

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer tokens = new StringTokenizer(reader.readLine());

        int n, m;
        n = Integer.parseInt(tokens.nextToken());
        m = Integer.parseInt(tokens.nextToken());

        for (int i = 0; i <= n; i++) {
            roads.add(new ArrayList<>());
        }
        for (int i = 0; i < m; i++) {
            tokens = new StringTokenizer(reader.readLine());
            int a, b;
            a = Integer.parseInt(tokens.nextToken());
            b = Integer.parseInt(tokens.nextToken());
            roads.get(a).add(new int[]{b, i});
            roads.get(b).add(new int[]{a, i});
        }

        colors = new String[m];
        Arrays.fill(colors, "G");

        for (int i = 1; i <= n; i++) {
            dfs(i, true);
        }

        System.out.println(String.join("", colors));
    }

    private static void dfs(int vertex, boolean useRed) {
        visited.add(vertex);
        for (int[] road : roads.get(vertex)) {
            if (!visited.contains(road[0])) {
                if (useRed) {
                    colors[road[1]] = "R";
                } else {
                    colors[road[1]] = "B";
                }
                dfs(road[0], !useRed);
            }
        }
    }
}

File: 2024/cpp/senior/senior4.cpp
#include <bits/stdc++.h>
using namespace std;

void dfs(vector<vector<pair<int, int>>> & roads, set<int> & visited, vector<char> & colors, int vertex, bool useRed) {
    visited.insert(vertex);
    for (auto road : roads[vertex]) {
        if (visited.find(road.first) == visited.end()) {
            if (useRed) {
                colors[road.second] = 'R';
            } else {
                colors[road.second] = 'B';
            }
            dfs(roads, visited, colors, road.first, !useRed);
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<vector<pair<int, int>>> roads(n + 1);
    vector<char> colors(m, 'G');
    set<int> visited{};

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        roads[a].emplace_back(b, i);
        roads[b].emplace_back(a, i);
    }

    for (int i = 1; i <= n; i++) {
        dfs(roads, visited, colors, i, true);
    }

    for (char c : colors) {
        cout << c;
    }
    cout << endl;
}
